[nmarwah7] iP

nmarwah7 opened the pull request.

Commit:
Level_0
Rename, Greet, Exit

Commit:
Add echo capability to chatbot

Chatbot cannot answer or interact with user.

Let's import scanner class from java.util and use it to read in a line
from user and print it back out in terminal.

Commit:
Add ability to store user input into task list and display it

There is no way for the chatbot to allow the user to input and create
any lists for functionality.

Let's add a String array of size 100 to store any tasks for the user,
and make use of "list" command to display these itemized tasks back to
user.

Commit:
Add ability to mark tasks as done or not

There is no way to keep the list updated with the current status of a
task. Task as a list of strings is too limited as a data type.

Let's keep tasks as a class on it's own and add isDone boolean value
to it, and add new commands to mark and unmark tasks. Add in a method
to handle out of bound demands to chatbot.

Commit:
Create javadocs comment for added method and clean up code to match
coding standard

Commit:
Add 3 new inherited classes for task and implement toString() for Task

Important to create inherited classes for different types of task class.
The print statements for tasks too long, and toString can be overridden
to optimize code.

The inherited classes are not implemented in main code logic yet.

Commit:
Add functional logic in Ultron.java for different task types

Create if-else extension to account for new task types to allow user
more flexibility

Commit:
Fix bug for user input for mark/unmark command and display for events

When user inputs mark -1, or index less than zero the program crashes
instead of displaying an out of bounds message. Creating an event task
had semantic [D] instead of [E].

Implement a check for taskNumber<0, and change toString of Event class

Commit:
Add task count when new task is added

User cannot see how many tasks have been added when creating a new task.

Add one line of code to print this after "added <task>:" string

Commit:
Use I/O technique to automate testing of Ultron

Testing involves retyping commands several times and manually observing
difference. This is tedious and error-prone.

Fix runtest.bat to automate testing, and update input and EXPECTED
files accordingly.

Commit:
Refactor code to improve code quality

Many lines of code and logic blocks are repetitive and hard to read.

Let's extract magic literals and repeated code as methods and variables
to try and achieve SLAP.

manavm12 commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+import java.util.Scanner;
+
+public class Ultron {
+    public static final String ULTRON_FACE = """
```
Naming of constants adheres to the coding standards

manavm12 commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+                    "But his idea of peace was... flawed.");
+        System.out.println("    Now, what do you need? ");
+        dashLine();
+    }
```
Naming is clear and understandable

manavm12 commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+            line = in.nextLine();
+        }
+        byeMessage();
+    }
```
Main function is too long. Code should be modular and broken down into more methods.

chwenyee commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+        dashLine();
+    }
+
+    /**
+     * Displays a standard error message if user tries to mark or unmark a task with index out of bound for the task list
+     */
```
I like this comment! It helps me to understand your outofBoundsMessage method better.

chwenyee commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+                }else {
+                    taskList[taskNumber].setDone(false);
+                    dashLine();
+                    System.out.println("    Moving backwards? How typical for humans.");
+                    System.out.println("    " + taskList[taskNumber]);
+                    dashLine();
+                }
+            }else if (line.contains("mark")) {
+                String stringTaskNumber = line.split(" ")[1];
+                int taskNumber = Integer.parseInt(stringTaskNumber)-1;
+                if(taskNumber>=Task.taskCount||taskNumber<0){
+                    outOfBoundsMessage();
+                }else {
+                    taskList[taskNumber].setDone(true);
+                    dashLine();
+                    System.out.println("    I hope you're not expecting a pat on the back. Marked done.");
+                    System.out.println("    " + taskList[taskNumber]);
+                    dashLine();
+                }
+            }else if(line.contains("todo")){
+                String todoDescription = line.split(" ",2)[1];
+                taskList[Task.taskCount] = new Todo(todoDescription);
+                taskAddedMessage(taskList, " todo ");
+            } else if(line.contains("deadline")){
+                String deadlineDescription = line.split("/by ")[0].split(" ")[1];
+                String deadlineBy = line.split("/by ")[1];
+                taskList[Task.taskCount] = new Deadline(deadlineDescription, deadlineBy);
+                taskAddedMessage(taskList, " deadline ");
+            } else if(line.contains("event")){
+                String eventDescription = line.split("/from ")[0].split(" ")[1];
+                String eventFrom = line.split("/from ")[1].split("/to ")[0];
+                String eventTo = line.split("/from ")[1].split("/to ")[1];
+                taskList[Task.taskCount] = new Event(eventDescription, eventFrom, eventTo);
+                taskAddedMessage(taskList, " event ");
+            } else { //this code should be redundant and removed
+                taskList[Task.taskCount] = new Task(line);
+                taskAddedMessage(taskList, " ");
+            }
+            line = in.nextLine();
+        }
```
Main method looks a bit long. Perhaps you could avoid it by separate different operations (such as unmark and mark) into several methods.

I see no deep nesting in your while loop. Good job.

chwenyee commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+                \\          /
+                 `--------`
+            """;
+    public static final int DASH_LINE_WIDTH = 120;
```
I noticed all your constant names follow the coding standard. Good job!

chwenyee commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+        System.out.println(ULTRON_FACE);
+        helloMessage();
+        Scanner in = new Scanner(System.in);
+        String line = in.nextLine();
```
Maybe you can add a blank line here to separate them into two groups to improve the readibility.

chwenyee commented on src/main/java/Ultron.java:
```diff
@@ -0,0 +1,117 @@
+
+    private static void taskAddedMessage(Task[] taskList, String taskType) {
+        dashLine();
+        //taskCount - 1 below to ensure null is not printed
```
This comment really helps me to understand the reason you add "-1" after taskCount for your implementation.

chwenyee reviewed (commented):
I think you did well overall! Good job on using methods to break down your main method as it really makes your code easier to read. The naming of classes and methods are clear too.

Commit:
Refactor code to switch-case and extract out common methods

Code is too lengthy in main() function and SLAP is not ideal for
readability.

Let's use tutorial PR exercise to help refactor and change the code
to improve readability for others. Includes a bug fix for lengthy
event and deadline descriptions.

Commit:
Add minimal error handling for empty todo and invalid command

Let's add minimal error checking for when todo description is empty,
and when invalid command is added. Additional check added for when
mark and unmark do not come with a valid task number.

Commit:
Add error handling for event and deadline commands

Invalid or empty strings can be entered by the user for deadline and
event command, which causes the code to crash. The main() function
has too many try-catch blocks, impairs readability of code.

Let's add in error-handling for empty or invalid command syntax for
user, and let's abstract out error-handling to the handle functions.

Commit:
Merge branch 'branch-Level-5'

Commit:
Add package structure to code base

All class files fall under ./java directory. This could get very
crowded.

Let's create package ultron, and subpackages ultron.tasks and
ultron.exceptions and move requisite files into each.

Commit:
Merge branch 'branch-A-Packages'

Commit:
Tweak runtest.bat to account for package structure

Commit:
Add collection interface for task list

Let's add a ArrayList<Task> for the tasklist to allow dynamic resizing
of the list and and change the subsequent references to it accordingly.

Commit:
Add in delete command functionality

Not possible to remove a task from the list.

Let's add a new command and method to handle deletion by user.

Commit:
Add save and read from file functionality

Commit:
Refactor out lengthy code to read in stored data

Commit:
Add update stored data method at end of main() method

This allows cleaner code and fewer preexisting methods to be altered.

Commit:
Cosmetic changes to task-added message and handleUnmark params

Commit:
Merge branch 'branch-Level-6'

Commit:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/ultron/Ultron.java

Commit:
Create an A-Jar file for first release

Commit:
Add in Ui class and refactor code to shift into new class

Let's extract all interaction methods to Ui class and refactor all the
code likewise by instantiating new Ui() class.

Commit:
Add storage and parser class and refactor code to match

Commit:
Add in Tasklist class and refactor to match, add in packages for all
refactored classes

Commit:
Add a filter by date and date formatting in Deadline tasks

Commit:
Add find command to allow user to search for keywords

No way to search for a certain task in a long list.

Let's add in a new method to Tasklist class to handle a find command by
creating a filtered list.

Commit:
Add error exception handling for the datefind method

Commit:
Add JavaDoc comments for non-trivial methods for documentation

nmarwah7 committed:
Merge pull request #3 from nmarwah7/branch-Level-9

Add find command to allow user to search for keywords

Commit:
Merge branch 'master' of https://github.com/nmarwah7/ip into branch-Level-8

# Conflicts:
#	src/main/java/ultron/Ultron.java
#	src/main/java/ultron/tasks/Tasklist.java

nmarwah7 committed:
Merge pull request #2 from nmarwah7/branch-Level-8

Branch level 8

Commit:
Merge branch 'master' of https://github.com/nmarwah7/ip into branch-A-JavaDoc

# Conflicts:
#	data/ultron.txt
#	src/main/java/ultron/Ultron.java
#	src/main/java/ultron/tasks/Deadline.java
#	src/main/java/ultron/tasks/Tasklist.java
#	src/main/java/ultron/ui/Ui.java

nmarwah7 committed:
Merge pull request #1 from nmarwah7/branch-A-JavaDoc

Add JavaDoc comments for non-trivial methods for documentation

Commit:
Merge branch 'master' of https://github.com/nmarwah7/ip

Commit:
Update README to include User Guide for Ultron in GFMD

Commit:
Update README to correct formatting of command summary table

The command summary table is improperly formatted.

Let's fix it according to GFMD convention

Commit:
Update README to correct formatting of command summary table to HTML

The command summary table is improperly formatted.

Let's fix it according to HTML convention.

Commit:
Refactor code to ensure javadoc comments are in the correct position

Add in a bug fix for deadline /by incorrect date parameters

Commit:
Change README to allow display of images for the webpage

Commit:
Refactor code in tasklist to fit OOP standards

Let's fix a few bugs regarding how packages are structured.

Commit:
Refactor code to fit OOP standards

Let's fix a few bugs regarding how packages are structured.

Commit:
Fix bug with datefind command and find command

Commit:
Update README to display image on GitHub Pages

Commit:
Add more images and improve clarity of README

Commit:
Change to README to update User Guide

okkhoy commented on src/main/java/ultron/exceptions/emptyCommandParameterException.java:
```diff
@@ -0,0 +1,6 @@
+/**
+ * Error thrown when a parameter expected for a command is left blank
+ */
+public class emptyCommandParameterException extends Exception{
```
class name begins with a capital letter, applies to all classes

okkhoy commented on src/main/java/ultron/parser/Parser.java:
```diff
@@ -0,0 +1,94 @@
+        return new Parser.EventParameters(eventDescription, eventFrom, eventTo);
+    }
+
+    public record EventParameters(String eventDescription, String eventFrom, String eventTo) {
```
empty method?

okkhoy commented on src/main/java/ultron/parser/Parser.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    public String getTodoParameters(String line) throws emptyCommandParameterException {
+        String todoDescription = line.split(" ",2)[1].trim();
+        if(todoDescription.trim().isEmpty()){
```
space between `){`

okkhoy commented on src/main/java/ultron/parser/Parser.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    public int getTaskNumber(String line) throws emptyCommandParameterException {
+        String stringTaskNumber = line.split(" ")[1];
+        if(stringTaskNumber.trim().isEmpty()){
```
spacing

okkhoy commented on src/main/java/ultron/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+ */
+public class Storage {
+
+    private static Ui ui = null; // Store Ui instance
```
redundant comment

okkhoy commented on src/main/java/ultron/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+                if(line.trim().isEmpty()){
+                    continue;
+                }
+                String taskType = line.split("|",2)[0];
+                String taskDescription = line.split("/description",2)[1];
+                Boolean isDoneTask = (!line.split("/done")[1]
```
consider moving low-level string operations to a separate method for better SLAP

okkhoy commented on src/main/java/ultron/tasks/Tasklist.java:
```diff
@@ -0,0 +1,286 @@
+        this.taskList = taskList;
+    }
+
+    public ArrayList<Task> taskList= new ArrayList<>();
+    /**
```
could use vertical spacing for better readability

okkhoy commented on src/main/java/ultron/tasks/Tasklist.java:
```diff
@@ -0,0 +1,286 @@
+            ArrayList<Task> filteredList = new ArrayList<>();
+            int filteredIndex = 0;
+            for (int i = 0; i < Task.taskCount; i++) {
+                if (taskList.get(i) instanceof Deadline) {
+                    if (((Deadline) taskList.get(i)).getBy().equals(date.trim())) {
+                        filteredList.add(taskList.get(i));
+                        filteredIndex++;
```
see how you can eliminate this arrow head code

okkhoy commented on src/main/java/ultron/ui/Ui.java:
```diff
@@ -0,0 +1,224 @@
+ * Contains methods related to interactions with the user on the front-end.
+ */
+public class Ui {
+    //the below face ASCII art was obtained through help of ChatGPT who was provided the prompt: "give me
+    //an Ultron face art in text" and the first output was used
```
👍

okkhoy commented on src/main/java/ultron/ui/Ui.java:
```diff
@@ -0,0 +1,224 @@
+                | \\IIIIII/ |
+                \\          /
+                 `--------`
+            """;
```
why does he look scared though 😄

okkhoy reviewed (commented):
@nmarwah7 
there are few coding standard violations. however, most can be fixed by using IDE help. consider fixing them before the final submission

Commit:
Tweak code to acceptable standards by PR feedback

Commit:
Tweak code to refine code quality
